### PR TITLE
Parser for the new types

### PR DESCRIPTION
**Pull request by konnov**
I will add more features to this PR later. But here is the first logical piece.

This PR contains a parser for Type System 1, as described in [ADR-002](https://github.com/informalsystems/apalache/blob/unstable/docs/adr/002adr-types.md).

### CODE REVIEW

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeParserFactory.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    */
+  def type1Parser(): Type1Parser = {
+    DefaultType1Parser
+  }
+
+  /**
+    * Create a caching parser for type system 1.
+    * @return a parser instance
+    */
+  def cacheType1Parser(): Type1Parser = {
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeParserFactory.scala:28]: Committing it separately, as it requires a bit of refactoring

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    * @return a type on success; throws TlcConfigParseError on failure
+    */
+  def apply(text: String): TlaType1 = {
+    apply(new StringReader(text))
+  }
+
+  /**
+    * Parse a type file from a reader.
+    * @param reader a reader
+    * @return a typeExpr on success; throws TlcConfigParseError on failure
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala:39]: also a separate commit

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+  }
+
+  // the access point
+  private def closedTypeExpr: Parser[TlaType1] = phrase(typeExpr) ^^ (e => e)
+
+  private def typeExpr: Parser[TlaType1] = {
+    // functions are tricky, as they start as other expressions, so we have to distinguish them by RIGHT_ARROW
+    (noFunExpr ~ opt((RIGHT_ARROW() | DOUBLE_RIGHT_ARROW()) ~ typeExpr)) ^^ {
+      case (lhs :: Nil) ~ Some(RIGHT_ARROW() ~ rt) => FunT1(lhs, rt)
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala:56]: it was meant to be "right type", but i will change it to `rhs` for consistency

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    * Parse the input stream and return the list of tokens. Although collecting the list of all tokens in memory is
+    * not optimal, TLC configurations files are tiny, so it should not be a big deal.
+    *
+    * @param reader a Java reader
+    * @return the list of parsed tokens
+    * @throws TlcConfigParseError when the lexer finds an error
+    */
+  def apply(reader: Reader): List[Type1Token] = parseAll(expr, reader) match {
+    case Success(result, _) => result
+    case NoSuccess(msg, next) => throw new Type1ParseError(msg, next.pos)
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala:26]: It is a very simple parser, so I don't feel that we should go very far here

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    positioned(
+      int | bool | str | set | seq | capsIdentifier | letterIdentifier | fieldIdentifier |
+        rightArrow | doubleRightArrow | leftParen | rightParen | leftBracket | rightBracket |
+        doubleLeftAngle | doubleRightAngle | comma | colon
+    ) ///
+
+  // a linefeed is not a white-space
+  def skip: Parser[Unit] = rep(whiteSpace) ^^^ Unit
+
+  private def capsIdentifier: Parser[CAPS_IDENT] = {
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala:43]: Its meant for the uninterpreted types, not for the record fields

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    * @throws TlcConfigParseError when the lexer finds an error
+    */
+  def apply(reader: Reader): List[Type1Token] = parseAll(expr, reader) match {
+    case Success(result, _) => result
+    case NoSuccess(msg, next) => throw new Type1ParseError(msg, next.pos)
+  }
+
+  def expr: Parser[List[Type1Token]] = skip ~> rep(token <~ skip) <~ eof
+
+  def eof: Parser[String] = "\\z".r | failure("unexpected character")
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala:31]: Good question. I have added a test for that. The parsing framework only gives us the position inside the string. I believe it is good enough, as the incorrect characters may be arbitrary.

**Comment by konnov** [comment]
I also felt that this PR contains lots of code already. So I will merge it and create and new branch afterward.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala:43]: Ah! My mistake. :+1:

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/Type1Lexer.scala:26]: Ok!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/parser/DefaultType1Parser.scala]: That works!

**Pull request merged**